supervisor/supervisor.go

```diff
@@ -117,10 +156,36 @@ func (s *Supervisor) worker() {
 		deleteEntries := make([]*sqs.DeleteMessageBatchRequestEntry, 0)
 		changeVisibilityEntries := make([]*sqs.ChangeMessageVisibilityBatchRequestEntry, 0)
 
+		batchTimes := make([]float64, 0, len(output.Messages))
 		for _, msg := range output.Messages {
+			// Perform a check on the message's validity and expire it if deemed too old.
+			if firstReceivedTSInSecondsStr, ok := msg.Attributes[firstRecvTimestampAttrKey]; ok && firstReceivedTSInSecondsStr != nil {
+				var firstReceivedTSInSeconds int64
+				if firstReceivedTSInSeconds, err = strconv.ParseInt(*firstReceivedTSInSecondsStr, 10, 64); err != nil {
+					s.logger.Errorf("Error while parsing first received timestamp (%s): %s", *firstReceivedTSInSecondsStr, err)
+				}
+				if nowUnix := time.Now().UTC().Unix(); nowUnix > firstReceivedTSInSeconds+int64(s.workerConfig.RetentionPeriodSeconds) {
+					sinceFirstSeen := nowUnix - firstReceivedTSInSeconds
+					s.logger.Debugf("Expired Message %d was first received %d seconds ago, deleting", msg.MessageId, sinceFirstSeen)
```

Comment:
What if instead of logging, we use this https://docs.aws.amazon.com/sdk-for-go/api/service/cloudwatch/#CloudWatch.PutMetricData and have a dashboard? Maybe an overkill, but should simplify this code. It would add the latency to the call though.

Reply:
we could do that in a coroutine, i can add it later for good observability